File: dijikstra's algo/1514.js
class MaxHeap{
    constructor(){
        this.heap =[]
    }
    push(value, priority){
        this.heap.push({value, priority})
        this.heap.sort((a, b) => b.priority - a.priority)
    }
    pop(){
        return this.heap.shift()
    }
    isEmpty(){
        return this.heap.length === 0
    }
}
function maxProbability(n, edges, succProb, start, end){
    const adj = Array.from({length: n}, () => [])
    edges.forEach(([u, v]) =>{
        adj[u].push([v, succProb[i]])
        adj[v].push([u, succProb[i]])
    })
    const result = new Array(n).fill(0.0)
    result[start] = 1.0
    const maxHeap = new MaxHeap()
    maxHeap.push(start, 1.0)
    while(!maxHeap.isEmpty()){
        const {value: currNode, priority: currProb} = maxHeap.pop()
        if(currNode === end) return currProb
        for(const [adjNode, adjProb] of adj[currNode]){
            const newProb = currProb * adjProb
            if(newProb > result[adjNode]){
                result[adjNode] = newProb
                maxHeap.push(adjNode, newProb)
            }
        }
    }
    return 0.0
}

File: dijikstra's algo/1514.py
# Google
# For probability needs to multiply
# In dijkstra we do sum but here just needs to multiply
# In this problem max probability instead minimum distance
# need to use max heap instead min heap
# Tc = O(N+E).logN
# Sc = O(N+E) [adj list and result vector]
import heapq
from collections import defaultdict

class Solution:
    def maxProbability(self, n: int, edges: list[list[int]], succProb: list[float], start: int, end: int) -> float:
        # Create an adjacency list
        adj = defaultdict(list)
        for i in range(len(edges)):
            u, v = edges[i]
            prob = succProb[i]
            adj[u].append((v, prob))
            adj[v].append((u, prob))
        
        # Max heap to store probabilities
        max_heap = []
        heapq.heappush(max_heap, (-1.0, start))  # Use negative to simulate max-heap
        result = [0.0] * n
        result[start] = 1.0  # Probability to reach start is 100%
        
        while max_heap:
            currProb, currNode = heapq.heappop(max_heap)
            currProb = -currProb  # Convert back to positive
            
            # If we reach the end node, return the probability
            if currNode == end:
                return currProb
            
            for adjNode, adjProb in adj[currNode]:
                newProb = currProb * adjProb
                if result[adjNode] < newProb:
                    result[adjNode] = newProb
                    heapq.heappush(max_heap, (-newProb, adjNode))  # Push negative for max-heap
        
        return 0.0  # If end is not reachable


